### PR TITLE
Changing note type to cloze while in edit mode bug fix

### LINKED ISSUE #11626: An error occurs when you change an ordinary word into a cloze
###### Reproduction Steps

1. Add a new word
2. Edit the word
3. Change the type to cloze
4. Fields cannot be edited
5. Previewing the word at this time will show an error

###### Expected Result

Fields can be edited normally

###### Actual Result

Fields cannot be edited, and previewing will fail

###### Debug info
AnkiDroid Version = 2.16alpha67

Android Version = 11

Manufacturer = OnePlus

Model = KB2000

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 11; KB2000 Build/RP1A.201005.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/99.0.4844.88 Mobile Safari/537.36

ACRA UUID = 5f8bfad0-e4da-46e7-a01a-753e09ec39d4

Scheduler = std2

Crash Reports Enabled = true

DatabaseV2 Enabled = true

###### Video

https://user-images.githubusercontent.com/101189845/173603520-dc9525c9-8ff7-4245-aea7-40212d3fa0ea.mp4

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by dorrin-sot**
## Purpose / Description
Changing note type to cloze while in edit mode would disable the `FieldEditLine`s and also saving a cloze-type note without any cloze deletions, wouldn't show error toast

## Fixes
Fixes #11626

## Approach
_How does this change address the problem?_

## How Has This Been Tested?

https://user-images.githubusercontent.com/59933477/173639764-e4bbc53f-5e04-4530-bf26-cd5c936280bd.mov

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 1]
```diff
@@ -724,15 +724,15 @@ class NoteEditor : AnkiActivity(), DeckSelectionListener, SubtitleListener, Tags
         }
         saveToggleStickyMap()
 
+        // Different from libAnki, block if there are no cloze deletions.
+        // DEFECT: This does not block addition if cloze transpositions are in non-cloze fields.
+        if (isClozeType && !hasClozeDeletions()) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: This was moved outside `mAddNote`: does this mean a user can't edit a cloze deletion note with no clozes (if imported from Anki Desktop)?

Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: No It's just that it prevents the user from saving a cloze note with no deletions (while in "Edit Note" mode), which can be if user has changed note type to cloze and not added any deletions or the user has edited an already cloze note and removed its deletions. Showing this error message and returning was the default behavior when a note was being **added** but it should also happen if a note is being **edited**

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: We have 3 states:

* new note (`mAddNote = true`)
* editing a note (`mAddNote = false`)
* changing the type of the note (remapping) (`mAddNote = false`)

I feel we're mistaking the 'remapping' state with the 'editing a note' state

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: Users currently aren't able to edit a card until it's been saved (if we are changing the note type, this lets users fix the field mapping).

This PR lets users edit the fields, but it disallows a user from setting the mapping if moving from a note type with 3 fields to a note type with 2 fields.

**Review by david-allison — CHANGES_REQUESTED** [review]
This removes the ability to map the fields when changing the note type.

I feel it'd be best to move the discussion back to the issue. 

We have a lot of UX tech debt around editing note types, and it'd be best to resolve this properly rather than adding more complexity.

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt (hunk 1 above)
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: But they can change note type type. they just can't save a cloze note with no deletions (which was previously the case for when user was creating a cloze note without deletions but not when editing note).

https://user-images.githubusercontent.com/59933477/174349048-05c2ec2f-3ba9-4853-a05e-73638693e3cf.mov

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: They can't change the field mapping of a note type. 

Change a note type with 4 fields to a note type with two, and you'll see what I mean

Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: Do you mean the last two fields getting cut off?

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: No, AnkiDroid changes the 'attachment' icon to a 'select field' icon

image

Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:729]: Oh sorry. So, if a user is changing model to "Cloze" this is the expected behaviour?
1- `FieldEditText` stay enabled
2- if number of fields is decreasing when model was changed, `editModelMode = true`
3- if user tries to save the note without any deletions, prevent

**Pull request closed without merging**